### PR TITLE
[RFC] `match_only_text` type migration - Stage 1

### PR DESCRIPTION
**Pull request by ebeahan**
### Summary

Continuing onto Stage 1 proposal for migrating `text` type fields, including the `message` base field, to use the `match_only_text` data type being introduced in Elasticsearch 7.14.

Stage 1 (Draft) Criteria:

- [x] Opened pull request for this proposal revising the existing strawperson
- [x] Identified "sponsor" at Elastic who will participate in RFC process and take ownership of the change after the process completes
- [x] Outlined initial field definitions
- [x] High-level description of examples of usage
- [x] High-level description of example sources of data
- [x] Identified potential concerns and implementation challenges/complexity
- [x] Subject matter experts identified and weighed in on the high level utility of these changes in the pull request
- [x] ECS team weighed in on appropriateness of these changes in the pull request
- [x] Assess potential performance impact

[Preview of markdown proposal](https://github.com/ebeahan/ecs/blob/rfc/0023/stage-1/rfcs/text/0023-match_only_text-data-type.md)

### CODE REVIEW

**Comment by ebeahan** [comment]
Notable changes for stage 1:

* I've extended the proposal also to include the current `text` multi-fields. I am happy to discuss more here if there are concerns or feedback, but I think we should pursue migrating to `match_only_text` across all of ECS as a replacement for `text`. For the majority of ECS use cases, users would likely benefit from the reduced storage use and not be impacted by the limitations of `match_only_text
* Listed the ECS team has a sponsor here. Since this would be a broad change across various field sets, I think it makes sense for ECS to sponsor the proposal. Again, open to discussion.
* Added detail to the `Scope of Impact`, `Source data`, and `Concerns` sections.

I'd appreciate it if reviewers would review and comment on:

* The list of candidate fields to migrate. Concerns with migrating all `text` fields?
* Any concerns with ECS team self-sponsoring?
* Any missing concerns or considerations?

File: rfcs/text/0023-match_only_text-data-type.md [hunk 1]
```diff
@@ -28,6 +24,60 @@ The following fields are currently indexed as `text` and are candidates to migra
 * `message`
 * `error.message`
 
+ECS also has `text` type multi-fields for several fields using the convention `<field-name>.text`. This RFC also proposes to migrate all `text` multi-fields to using `match_only_text`, but the multi-field will remain using the `<field-name>.text` to avoid a breaking change.
```
Comment (jpountz) [inline rfcs/text/0023-match_only_text-data-type.md:27]: I'd drop the mention to a breaking change. Say we were starting ECS from scratch today, I don't think we would be considering calling the sub field `match_only_text`?

File: rfcs/text/0023-match_only_text-data-type.md [hunk 2]
```diff
@@ -28,6 +24,60 @@ The following fields are currently indexed as `text` and are candidates to migra
...
+
+As part of this migration,
```
Comment (jpountz) [inline rfcs/text/0023-match_only_text-data-type.md:29]: I was expecting an end to this sentence after the list, is it missing?

File: rfcs/text/0023-match_only_text-data-type.md [hunk 3]
```diff
@@ -74,6 +124,105 @@ Stage 2: Included a real world example source document. Ideally this example com
...
+    "ecs_version": "2.0.0-dev"
+  },
+  "template": {
+    "mappings": {
+      "properties": {
+        "error": {
+          "properties": {
+            "message": {
+              "type": "match_only_text"
+            },
```
Comment (jpountz) [inline rfcs/text/0023-match_only_text-data-type.md:164]:
```suggestion
            }
```

File: rfcs/text/0023-match_only_text-data-type.md [hunk 4]
```diff
@@ -84,12 +233,33 @@ Stage 2: Identifies scope of impact of changes. Are breaking changes required? S
...
+Ingestion mechanisms will need to adopt the new mappings to index fields as the new type. Since both `text` and `match_only_text` are members of the `text` type family, this will not cause a conflict in Kibana index patterns across indices using the two types for the same field.
+
+As with all type changes in ECS, the ECS team will benchmark and identify any significant storage, indexing, or query changes. The Elasticsearch performance will be engaged, if necessary, to determine the results. However, `match_only_text` is a `text` field with these settings:
+
+* `index_options: docs`
+* `norms: false`
+
+As a convention, ECS already sets `norms: false` on all `text` fields, so this setting should have no impact on performance. Setting `index_options: docs` only indexes the doc number and has already been recommended to reduce the disk usage needed for indexing.
+
+Negative performance or storage side-effects from this change are not expected.
```
Comment (jpountz) [inline rfcs/text/0023-match_only_text-data-type.md:245]: Except for phrase and intervals queries. I know it's stated a couple lines below, but taken out of context this sentence could suggest that everything is faster or as fast.

File: rfcs/text/0023-match_only_text-data-type.md [hunk 5]
```diff
@@ -28,6 +24,60 @@ The following fields are currently indexed as `text` and are candidates to migra
...
+* `user.effective.full_name.text`
+* `user.effective.name.text`
+* `user.full_name.text`
+* `user.name.text`
+* `user.target.full_name.text`
+* `user.target.name.text`
+* `user_agent.original.text`
+* `user_agent.os.full.text`
+* `user_agent.os.name.text`
+* `vulnerability.description.text`
```
Comment (jpountz) [inline rfcs/text/0023-match_only_text-data-type.md:80]: Out of curiosity, why are we listing all multi-field usage, and no direct usage such as `message` and `error.message`?

File: rfcs/text/0023-match_only_text-data-type.md (hunk 2 above)
Comment (ebeahan) [reply rfcs/text/0023-match_only_text-data-type.md]: No, I missed it during editing. Removed.

File: rfcs/text/0023-match_only_text-data-type.md (hunk 1 above)
Comment (ebeahan) [reply rfcs/text/0023-match_only_text-data-type.md]: Good point - removed mention.

File: rfcs/text/0023-match_only_text-data-type.md (hunk 4 above)
Comment (ebeahan) [reply rfcs/text/0023-match_only_text-data-type.md]: Rephrased this line to note this as a known limitation.

File: rfcs/text/0023-match_only_text-data-type.md (hunk 5 above)
Comment (ebeahan) [reply rfcs/text/0023-match_only_text-data-type.md:82]: They are listed above the multi-fields: https://github.com/ebeahan/ecs/blob/rfc/0023/stage-1/rfcs/text/0023-match_only_text-data-type.md#fields.

I've added section headers to distinguish the two sections.

**Review by jpountz — APPROVED** [review]
:+1:

**Review by djptek — COMMENTED** [review]
Suggest to add actionable item for performance impact analysis on both the two original fields as well as the 49 multi-fields added in Stage 1

**Comment by djptek** [comment]
sent a merge so the checks would pass

**Comment by ebeahan** [comment]
After some out-of-band discussion on performance testing these changes, I'm removing the `.text` multi-fields from this RFC. This will narrow the scope of the proposed change to two `text` fields: `message` and `error.message`.

 Migrating the `.text` multi-fields will be proposed in a future RFC (and perhaps multiple future RFCs).

**Review by djptek — APPROVED** [review]
LGTM

**Comment by ebeahan** [comment]
Holding off on merging until potential performance impact is assessed.

**Pull request merged**